From 994fc3d04fc61a1bcc3d57f6b0da08ef2ef053e6 Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Mon, 3 Jul 2023 16:14:17 +0800
Subject: [PATCH 612/780] [Huawei] PSI: add more memory fine grained stall
 tracking in pressure.stat

Offering: HULK
HULK inclusion
category: feature
bugzilla: 188780

-------------------------------

Introcude more memory fine grianed stall tracking in pressure.stat, such
as global memory relcaim, memory compact, memory async cgroup reclaim
and swap.

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 include/linux/psi_types.h | 24 +++++++++++++++++++++
 kernel/sched/psi.c        | 44 +++++++++++++++++++++++++++++++++++++++
 mm/memcontrol.c           |  1 +
 mm/page_alloc.c           |  2 ++
 mm/page_io.c              |  1 +
 mm/vmscan.c               |  1 +
 6 files changed, 73 insertions(+)

diff --git a/include/linux/psi_types.h b/include/linux/psi_types.h
index 4fc6b47d3f7f..1224951694ef 100644
--- a/include/linux/psi_types.h
+++ b/include/linux/psi_types.h
@@ -233,12 +233,28 @@ struct psi_group {
 enum psi_stat_states {
 	PSI_MEMCG_RECLAIM_SOME,
 	PSI_MEMCG_RECLAIM_FULL,
+	PSI_GLOBAL_RECLAIM_SOME,
+	PSI_GLOBAL_RECLAIM_FULL,
+	PSI_COMPACT_SOME,
+	PSI_COMPACT_FULL,
+	PSI_ASYNC_MEMCG_RECLAIM_SOME,
+	PSI_ASYNC_MEMCG_RECLAIM_FULL,
+	PSI_SWAP_SOME,
+	PSI_SWAP_FULL,
 	NR_PSI_STAT_STATES,
 };
 
 enum psi_stat_task_count {
 	NR_MEMCG_RECLAIM,
 	NR_MEMCG_RECLAIM_RUNNING,
+	NR_GLOBAL_RECLAIM,
+	NR_GLOBAL_RECLAIM_RUNNING,
+	NR_COMPACT,
+	NR_COMPACT_RUNNING,
+	NR_ASYNC_MEMCG_RECLAIM,
+	NR_ASYNC_MEMCG_RECLAIM_RUNNING,
+	NR_SWAP,
+	NR_SWAP_RUNNING,
 	NR_PSI_STAT_TASK_COUNTS,
 };
 
@@ -253,6 +269,10 @@ enum psi_stat_task_count {
  */
 enum psi_memstall_type {
 	PSI_MEMCG_RECLAIM = 1,
+	PSI_GLOBAL_RECLAIM,
+	PSI_COMPACT,
+	PSI_ASYNC_MEMCG_RECLAIM,
+	PSI_SWAP,
 };
 
 struct psi_group_stat_cpu {
@@ -277,6 +297,10 @@ struct psi_group_ext {
 
 struct psi_group_ext { };
 #define PSI_MEMCG_RECLAIM		0
+#define PSI_GLOBAL_RECLAIM		0
+#define PSI_COMPACT			0
+#define PSI_ASYNC_MEMCG_RECLAIM		0
+#define PSI_SWAP			0
 
 #endif /* CONFIG_PSI_FINE_GRAINED */
 
diff --git a/kernel/sched/psi.c b/kernel/sched/psi.c
index ca78a03b86f2..c96ac0737397 100644
--- a/kernel/sched/psi.c
+++ b/kernel/sched/psi.c
@@ -370,6 +370,26 @@ static void record_stat_times(struct psi_group_ext *psi_ext, int cpu)
 		if (ext_grpc->state_mask & (1 << PSI_MEMCG_RECLAIM_FULL))
 			ext_grpc->times[PSI_MEMCG_RECLAIM_FULL] += delta;
 	}
+	if (ext_grpc->state_mask & (1 << PSI_GLOBAL_RECLAIM_SOME)) {
+		ext_grpc->times[PSI_GLOBAL_RECLAIM_SOME] += delta;
+		if (ext_grpc->state_mask & (1 << PSI_GLOBAL_RECLAIM_FULL))
+			ext_grpc->times[PSI_GLOBAL_RECLAIM_FULL] += delta;
+	}
+	if (ext_grpc->state_mask & (1 << PSI_COMPACT_SOME)) {
+		ext_grpc->times[PSI_COMPACT_SOME] += delta;
+		if (ext_grpc->state_mask & (1 << PSI_COMPACT_FULL))
+			ext_grpc->times[PSI_COMPACT_FULL] += delta;
+	}
+	if (ext_grpc->state_mask & (1 << PSI_ASYNC_MEMCG_RECLAIM_SOME)) {
+		ext_grpc->times[PSI_ASYNC_MEMCG_RECLAIM_SOME] += delta;
+		if (ext_grpc->state_mask & (1 << PSI_ASYNC_MEMCG_RECLAIM_FULL))
+			ext_grpc->times[PSI_ASYNC_MEMCG_RECLAIM_FULL] += delta;
+	}
+	if (ext_grpc->state_mask & (1 << PSI_SWAP_SOME)) {
+		ext_grpc->times[PSI_SWAP_SOME] += delta;
+		if (ext_grpc->state_mask & (1 << PSI_SWAP_FULL))
+			ext_grpc->times[PSI_SWAP_FULL] += delta;
+	}
 }
 
 static bool test_fine_grained_stat(unsigned int *stat_tasks,
@@ -382,6 +402,26 @@ static bool test_fine_grained_stat(unsigned int *stat_tasks,
 	case PSI_MEMCG_RECLAIM_FULL:
 		return unlikely(stat_tasks[NR_MEMCG_RECLAIM] &&
 		       nr_running == stat_tasks[NR_MEMCG_RECLAIM_RUNNING]);
+	case PSI_GLOBAL_RECLAIM_SOME:
+		return unlikely(stat_tasks[NR_GLOBAL_RECLAIM]);
+	case PSI_GLOBAL_RECLAIM_FULL:
+		return unlikely(stat_tasks[NR_GLOBAL_RECLAIM] &&
+		       nr_running == stat_tasks[NR_GLOBAL_RECLAIM_RUNNING]);
+	case PSI_COMPACT_SOME:
+		return unlikely(stat_tasks[NR_COMPACT]);
+	case PSI_COMPACT_FULL:
+		return unlikely(stat_tasks[NR_COMPACT] &&
+		       nr_running == stat_tasks[NR_COMPACT_RUNNING]);
+	case PSI_ASYNC_MEMCG_RECLAIM_SOME:
+		return unlikely(stat_tasks[NR_ASYNC_MEMCG_RECLAIM]);
+	case PSI_ASYNC_MEMCG_RECLAIM_FULL:
+		return unlikely(stat_tasks[NR_ASYNC_MEMCG_RECLAIM] &&
+		      nr_running == stat_tasks[NR_ASYNC_MEMCG_RECLAIM_RUNNING]);
+	case PSI_SWAP_SOME:
+		return unlikely(stat_tasks[NR_SWAP]);
+	case PSI_SWAP_FULL:
+		return unlikely(stat_tasks[NR_SWAP] &&
+		       nr_running == stat_tasks[NR_SWAP_RUNNING]);
 	default:
 		return false;
 	}
@@ -1711,6 +1751,10 @@ static const struct proc_ops psi_cpu_proc_ops = {
 #ifdef CONFIG_PSI_FINE_GRAINED
 static const char *const psi_stat_names[] = {
 	"cgroup_memory_reclaim",
+	"global_memory_reclaim",
+	"compact",
+	"cgroup_async_memory_reclaim",
+	"swap",
 };
 
 int psi_stat_show(struct seq_file *m, struct psi_group *group)
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 870af5412663..28d0ac5ced6c 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -2397,6 +2397,7 @@ static void async_reclaim_high(struct mem_cgroup *memcg)
 		return;
 	}
 
+	psi_memstall_type_change(PSI_ASYNC_MEMCG_RECLAIM);
 	psi_memstall_enter(&pflags);
 	nr_pages = memcg_usage > safe_pages ? memcg_usage - safe_pages :
 		   MEMCG_CHARGE_BATCH;
diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index 2145a67894b7..ade657be9b5a 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -4180,6 +4180,7 @@ __alloc_pages_direct_compact(gfp_t gfp_mask, unsigned int order,
 	if (!order)
 		return NULL;
 
+	psi_memstall_type_change(PSI_COMPACT);
 	psi_memstall_enter(&pflags);
 	noreclaim_flag = memalloc_noreclaim_save();
 
@@ -4425,6 +4426,7 @@ __perform_reclaim(gfp_t gfp_mask, unsigned int order,
 
 	/* We now go into synchronous reclaim */
 	cpuset_memory_pressure_bump();
+	psi_memstall_type_change(PSI_GLOBAL_RECLAIM);
 	psi_memstall_enter(&pflags);
 	fs_reclaim_acquire(gfp_mask);
 	noreclaim_flag = memalloc_noreclaim_save();
diff --git a/mm/page_io.c b/mm/page_io.c
index ee28c39e566e..4a67dfcb618d 100644
--- a/mm/page_io.c
+++ b/mm/page_io.c
@@ -341,6 +341,7 @@ int swap_readpage(struct page *page, bool synchronous)
 	 * or the submitting cgroup IO-throttled, submission can be a
 	 * significant part of overall IO time.
 	 */
+	psi_memstall_type_change(PSI_SWAP);
 	psi_memstall_enter(&pflags);
 
 	if (frontswap_load(page) == 0) {
diff --git a/mm/vmscan.c b/mm/vmscan.c
index a390bd15fdea..e7e06a8624d8 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -4409,6 +4409,7 @@ static int __node_reclaim(struct pglist_data *pgdat, gfp_t gfp_mask, unsigned in
 					   sc.gfp_mask);
 
 	cond_resched();
+	psi_memstall_type_change(PSI_GLOBAL_RECLAIM);
 	psi_memstall_enter(&pflags);
 	fs_reclaim_acquire(sc.gfp_mask);
 	/*
-- 
2.17.1

